### DISCUSSION TITLE
Is there a way to get the optimal objective value of MILP solved by SCIP as information_function?

### DISCUSSION THREAD

**Discussion by thupchnsky** [ds4dm/ecole#164] [Questions] [answered]
Hi,

I know that there are many things in ecole.reward can be added into information function when defining the environment. For example, I can define a SCIP solver as `default_env = ecole.environment.Configuring(observation_function=None, information_function={"nb_nodes": ecole.reward.NNodes(), "time": ecole.reward.SolvingTime()}, scip_params=scip_parameters)`. So I can get the number of nodes explored and the solving time. However, is there a reward function that returns the objective value of the original optimization problem? I try to search it in doc but could not find it.

**Comment by AntoinePrv** [comment, marked as answer]
Hello @thupchnsky 

There is no such things in Ecole for now, but you can define it for yourself using PySCIPOpt:

```python
class BestObjVal:
    def before_reset(self, model):  # Edit added missing function
        pass

    def extract(self, model, done):
        m = model.as_pyscipopt()
        return m.getSolObjVal(m.getBestSol())
```
And then:
```python
default_env = ecole.environment.Configuring(
    observation_function=None,
    information_function={
        "nb_nodes": ecole.reward.NNodes(),
        "time": ecole.reward.SolvingTime(),
        "best_obj": BestObjVal(),
    },
    scip_params=scip_parameters
)
```

**Reply by thupchnsky** [reply]
Thanks for your reply! It seems that the custom class must have a reset method. Otherwise it gives me errors. However, when I add a dummy reset function to `BestObjVal` (I guess nothing needs to be reset here?..), it will cause fatal error to my jupyter kernel and force it to restart automatically. So I am wondering what should be the correct reset method here? Thanks!

**Reply by AntoinePrv** [reply]
> It seems that the custom class must have a reset method. Otherwise it gives me errors. 

You're right, my bad. An empty `before_reset` is needed.

> When I add a dummy reset function to BestObjVal (I guess nothing needs to be reset here?..), it will cause fatal error

It seems the problem is coming PySCIPOpt.

In `Configuring`, in `reset`, nothing has been done so SCIP is in Stage `SCIP_STAGE_PREOBLEM`.
In this case, [`SCIPgetBestSol`](https://www.scipopt.org/doc-7.0.2/html/group__PublicSolutionMethods.php#ga9eeada1341dee4fc47f84e93f5ea02ac) may return a `NULL` pointer because no solution exists, then in `getSolObjVal` `pyscipopt` does not check that `NULL` pointer and crashes.

Fortunaltely, `getNSols` does not do that, therefore, we add a check:
```python
def extract(self, model, done):
    m = model.as_pyscipopt()
    if m.getNSols() <= 0:
        return ...  # decide what to do
    else:
        return m.getSolObjVal(m.getBestSol())
```

**Reply by thupchnsky** [reply]
It works! Thanks for the reply!

**Reply by AntoinePrv** [reply]
For reference, I opened an issue https://github.com/scipopt/PySCIPOpt/issues/493
